community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/FilterFunction.scala

```diff
@@ -26,16 +26,24 @@ import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.QueryState
 import org.neo4j.values.AnyValue
 import org.neo4j.values.virtual.VirtualValues
 
+import scala.collection.mutable.ArrayBuffer
+
 case class FilterFunction(collection: Expression, id: String, predicate: Predicate)
   extends NullInNullOutExpression(collection)
   with ListSupport
   with Closure {
 
   override def compute(value: AnyValue, m: ExecutionContext, state: QueryState) = {
-    val traversable = makeTraversable(value)
-    VirtualValues.filter(traversable, new java.util.function.Function[AnyValue, java.lang.Boolean]() {
-      override def apply(v1: AnyValue): java.lang.Boolean =  predicate.isTrue(m.newWith1(id, v1), state)
-    })
+    val list = makeTraversable(value)
+    val filtered = new ArrayBuffer[AnyValue]
+    val inputs = list.iterator()
```

Comment:
Same here and since we know longer have filtered lists we are sure that we have fast random access to lists. So we can get rid of one object allocation here and use `list.value(i)` below instead